### PR TITLE
Added TypeScript definitions for firebase.functions()

### LINKED ISSUE #705: Add TypeScript definitions for firebase.functions()
* Operating System version: All
  * Firebase SDK version: 4.12.1
  * Firebase Product: functions

In this version there are no TypeScript definitions for functions. Resulting in a ts error of:

`[ts] Property 'functions' does not exist on type 'typeof firebase'.`

### PR DESCRIPTION
**Pull request by mpoehler**
Fixes #705, "Add TypeScript definitions for firebase.functions()", also described here: https://stackoverflow.com/questions/49669360/using-firebase-functions-from-web-sdk-using-typescript

### CODE REVIEW

File: packages/firebase/index.d.ts [hunk 1]
```diff
@@ -123,6 +127,42 @@ declare namespace firebase.app {
...
+    readonly data: any;
+  }
+  export interface HttpsCallable {
+    (data?: any): Promise<HttpsCallableResult>;
+  }
+  export class FirebaseFunctions {
+    private constructor();
+    httpsCallable(name: string): HttpsCallable;
+  }
+  export type FunctionsErrorCode =
```
Comment (bklimt) [inline packages/firebase/index.d.ts]: This should be named `ErrorStatus`.

**Comment by mpoehler** [comment]
In addition to the changes mentioned, I renamed the FirestoreErrorCodes to ErrorStatus in the firestore namespace to be consistent.

**Comment by jshcrowthe** [comment]
If we can roll back the `FirestoreErrorCodes` change I think this is good to go. Changing the name of an exported type (and not maintaining the previous name) is a breaking change, and I'd like to avoid that in this PR.

**Comment by bklimt** [comment]
Sorry, my mistake. I agree with jshcrowthe@ that we renaming it isn't worth making a breaking change.

File: packages/firebase/index.d.ts [hunk 2]
```diff
@@ -1934,7 +1972,7 @@ declare namespace firebase.firestore {
    * - 'unauthenticated': The request does not have valid authentication
    *   credentials for the operation.
    */
-  export type FirestoreErrorCode =
+  export type ErrorStatus =
```
Comment (jshcrowthe) [inline packages/firebase/index.d.ts]: Let's keep this change scoped to just the Functions SDK for now.

**Comment by mpoehler** [comment]
@jshcrowthe I returned back to FirestoreErrorStatus.

**Comment by jshcrowthe** [comment]
Awesome! Thanks for your help @mpoehler

**Pull request merged**